include/aws/common/cache.h

```diff
@@ -0,0 +1,92 @@
+#ifndef AWS_COMMON_CACHE_H
+#define AWS_COMMON_CACHE_H
+/*
+ * Copyright 2010-2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/linked_hash_table.h>
+
+struct aws_cache;
+
+struct aws_cache_vtable {
+    int (*find)(struct aws_cache *cache, const void *key, void **p_value);
+    int (*put)(struct aws_cache *cache, const void *key, void *p_value);
+    int (*remove)(struct aws_cache *cache, const void *key);
+    void (*clear)(struct aws_cache *cache);
+    size_t (*get_element_count)(const struct aws_cache *cache);
+};
```

Comment:
Also need a `void (*destroy)(struct aws_cache *cache)` entry.
Otherwise, there's no way for a concrete type to free anything in `impl`.